pkgs/build-support/cc-wrapper/default.nix

```diff
@@ -347,6 +355,8 @@ stdenv.mkDerivation {
       touch "$out/nix-support/libc-cflags"
       touch "$out/nix-support/libc-ldflags"
       echo "-B${libc_lib}${libc.libdir or "/lib/"}" >> $out/nix-support/libc-crt1-cflags
+    '' + optionalString includeFortifyHeaders' ''
+      echo "-isystem ${fortify-headers}/include" >> $out/nix-support/libc-cflags
```

Comment:
Silly question: why `-isystem` and not `-idirafter` like other entries in this file?

Reply:
The block comment a little bit above explains it better. Basically, the uses of `-idirafter` here are trying to supply a set of headers which will be referred to _after_ libc's, but we need to inject `fortify-headers` _before_ libc's.

Though it's a good point I shouldn't put my block between the `-B` addition and the `-idirafter` addition it necessitates.

Reply:
I'm not sure this comment is accurate WRT `-B` handling (AFAIU `-B` only redirects lookup of `.o` and executable tools search). `-B` should not affect include search.

It also does not touch on `-idirafter` vs `-isystem` (or I fail to read it as such). My understanding is that if you want two paths for libc headers you just use two identical options: `-I pre -I post`, or `-idirafter pre -idirafter post`. At least you get them sequentially. If you are using `-isystem pre -idirafter post` you get a bap between them where other stuff can get in (apart from special duplicates `-I` handling).

I think `-isystem` inserts headers before `gcc`'s default headers. `-idirafter` inserts after default `gcc`'s headers. I would say the canonical order is:
1. user's `-I`
2. `gcc`'s own includes
3. libc includes

`nixpkgs` usually abuses `-isystem` to insert c++ headers lookup path early (earlier than `gcc`'s own "fixed" or wrapping headers around libc, only then libc goes). You probably wan to insert `fortify-headers` after [2.], not after [1.] as it will override `gcc`'s headers.

The annoyance though is that `musl` does not match `glibc` in this picture and slightly messes up "`gcc`'s own includes" step and makes headers order even more confusing: https://github.com/NixOS/nixpkgs/issues/142066

Thus maybe it's a bit easier to reason about in a concrete change (intended versus actual). This is the search path before your change:

```
$ nix build -L --impure --expr 'with import ./. {}; pkgsMusl.stdenv.mkDerivation { name = "probe"; buildCommand = "echo | g++ -x c++ -E -v - |& fgrep -B10 search; exit 1"; }' |& unnix

probe> #include "..." search starts here:
probe> #include <...> search starts here:
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1/x86_64-unknown-linux-musl
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1/backward
probe>  /<<NIX>>/gcc-13.0.0/include
probe>  /<<NIX>>/musl-1.2.3-dev/include
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include-fixed
```

This is the search after your change:

```
$ nix build -L --impure --expr 'with import ./. {}; pkgsMusl.stdenv.mkDerivation { name = "probe"; buildCommand = "echo | g++ -x c++ -E -v - |& fgrep -B10 search; exit 1"; }' |& unnix

probe>  /<<NIX>>/fortify-headers-1.1alpine1/include
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1/x86_64-unknown-linux-musl
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../include/c++/13.0.1/backward
probe>  /<<NIX>>/gcc-13.0.0/include
probe>  /<<NIX>>/musl-1.2.3-dev/include
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include
probe>  /<<NIX>>/gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include-fixed
```

Note the distance between `fortify-headers` and `musl-1.2.3-dev/include`. I would say ideally they should go in order one right after another.

Reply:
I did of course start out using `-idirafter`, but for me at least, building `pkgsStatic.tests.hardeningFlags.fortify1ExplicitEnabledExecTest` from #217390 (with gcc'c `-v` flag hacked in to its `writeCBinWithStdenv`) gives me:

```
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/???-x86_64-unknown-linux-musl-stage-final-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/../../../../x86_64-unknown-linux-musl/sys-include
 /nix/store/???-x86_64-unknown-linux-musl-stage-final-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/../../../../x86_64-unknown-linux-musl/include
 /nix/store/???-x86_64-unknown-linux-musl-stage-final-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include
 /nix/store/???-musl-static-x86_64-unknown-linux-musl-1.2.3-dev/include
 /nix/store/???-x86_64-unknown-linux-musl-stage-final-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include-fixed
End of search list.
```

with gcc's `sys-include` dir having all the c standard headers, so this didn't work.

Reply:
Ah, these cross compilers! I think they are genuinely broken in this regard. The full include list is:

```
$ nix build -L --impure --expr 'with import ./. {}; pkgsStatic.stdenv.mkDerivation { name = "probe"; buildCommand = "echo | $CXX -x c++ -E -v - |& fgrep -B10 search; exit 1"; }' |& unnix

#include "..." search starts here:
#include <...> search starts here:
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../x86_64-unknown-linux-musl/include/c++/13.0.1
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../x86_64-unknown-linux-musl/include/c++/13.0.1/x86_64-unknown-linux-musl
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../x86_64-unknown-linux-musl/include/c++/13.0.1/backward
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../x86_64-unknown-linux-musl/sys-include
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/../../../../x86_64-unknown-linux-musl/include
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include
 /<<NIX>>/musl-static-x86_64-unknown-linux-musl-1.2.3-dev/include
 /<<NIX>>/x86_64-unknown-linux-musl-stage-final-gcc-13.0.0/lib/gcc/x86_64-unknown-linux-musl/13.0.1/include-fixed
probe-x86_64-unknown-linux-musl> End of search list.
```

Here `sys-include` completely shadows `/<<NIX>>/musl-static-x86_64-unknown-linux-musl-1.2.3-dev/include` injected later. I think it's a major problem as we effectively can't replace libc. I'll try to send a PR against `gcc` that removes it.

Reply:
Proposed removal of `sys-include` as https://github.com/NixOS/nixpkgs/pull/219811

Reply:
But a similar thing happens with the `pkgsMusl` variant if I use `-idirafter` instead (based on 0a575eb with the line in its original place):

```
extra flags before to /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/bin/gcc:
  -pie
  -fPIE
  -fPIC
  -Wformat
  -Wformat-security
  -Werror=format-security
  -fstack-protector-strong
  --param
  ssp-buffer-size=4
  -O2
  -U_FORTIFY_SOURCE
  -D_FORTIFY_SOURCE=2
  -fno-strict-overflow
  -Wl\,-dynamic-linker=/nix/store/kzf367gnfsxlk1siars8347dljqqkpmz-musl-1.2.3/lib/ld-musl-x86_64.so.1
original flags to /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/bin/gcc:
  -v
  -x
  c
  code.c
  -O1
  -o
  /nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/bin/test-bin
extra flags after to /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/bin/gcc:
  -B/nix/store/kzf367gnfsxlk1siars8347dljqqkpmz-musl-1.2.3/lib/
  -idirafter
  /nix/store/db2ifa4kihay3mvi5j5239sj4m7nj5jm-fortify-headers-1.1alpine1/include
  -idirafter
  /nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include
  -idirafter
  /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include-fixed
  -B/nix/store/i8xab269w7xykjgg4f6xj379v9mq2wgx-gcc-12.2.0-lib/lib
  -B/nix/store/wj9bdrb498zqvrcijkgwj40wc3rwjbjj-gcc-wrapper-12.2.0/bin/
  -frandom-seed=f4ygis0k6k
  -Wl\,-rpath
  -Wl\,/nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/lib64
  -Wl\,-rpath
  -Wl\,/nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/lib
  -L/nix/store/kzf367gnfsxlk1siars8347dljqqkpmz-musl-1.2.3/lib
  -L/nix/store/i8xab269w7xykjgg4f6xj379v9mq2wgx-gcc-12.2.0-lib/lib
Using built-in specs.
COLLECT_GCC=/nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/bin/gcc
COLLECT_LTO_WRAPPER=/nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/libexec/gcc/x86_64-unknown-linux-musl/12.2.0/lto-wrapper
Target: x86_64-unknown-linux-musl
Configured with:
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 12.2.0 (GCC)
COLLECT_GCC_OPTIONS='-pie' '-fPIC' '-Wformat=1' '-Wformat-security' '-Werror=format-security' '-fstack-protector-strong' '--param=ssp-buffer-size=4' '-O2' '-U' '_FORTIFY_SOURCE' '-D' '_FORTIFY_SOURCE=2' '-fno-strict-overflow' '-v' '-O1' '-o' '/nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/bin/test-bin' '-B' '/nix/store/kzf3
67gnfsxlk1siars8347dljqqkpmz-musl-1.2.3/lib/' '-idirafter' '/nix/store/db2ifa4kihay3mvi5j5239sj4m7nj5jm-fortify-headers-1.1alpine1/include' '-idirafter' '/nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include' '-idirafter' '/nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0
/include-fixed' '-B' '/nix/store/i8xab269w7xykjgg4f6xj379v9mq2wgx-gcc-12.2.0-lib/lib' '-B' '/nix/store/wj9bdrb498zqvrcijkgwj40wc3rwjbjj-gcc-wrapper-12.2.0/bin/' '-frandom-seed=f4ygis0k6k' '-L/nix/store/kzf367gnfsxlk1siars8347dljqqkpmz-musl-1.2.3/lib' '-L/nix/store/i8xab269w7xykjgg4f6xj379v9mq2wgx-gcc-12.2.0-lib/lib' '-mtune=gener
ic' '-march=x86-64' '-dumpdir' '/nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/bin/test-bin-'
 /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/libexec/gcc/x86_64-unknown-linux-musl/12.2.0/cc1 -quiet -v -U _FORTIFY_SOURCE -D _FORTIFY_SOURCE=2 -idirafter /nix/store/db2ifa4kihay3mvi5j5239sj4m7nj5jm-fortify-headers-1.1alpine1/include -idirafter /nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include -idiraft
er /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include-fixed code.c -quiet -dumpdir /nix/store/f4ygis0k6kdmycdaps74da05wj1j8x0s-test-bin/bin/test-bin- -dumpbase code.c -dumpbase-ext .c -mtune=generic -march=x86-64 -O2 -O1 -Wformat=1 -Wformat-security -Werror=format-security -ver
sion -fPIC -fstack-protector-strong -fno-strict-overflow -frandom-seed=f4ygis0k6k --param=ssp-buffer-size=4 -o /build/ccMijLpJ.s
GNU C17 (GCC) version 12.2.0 (x86_64-unknown-linux-musl)
        compiled by GNU C version 12.2.0, GMP version 6.2.1, MPFR version 4.2.0, MPC version 1.3.1, isl version isl-0.20-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/../../../../x86_64-unknown-linux-musl/include"
ignoring duplicate directory "/nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include"
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/include
 /nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include
 /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include
 /nix/store/db2ifa4kihay3mvi5j5239sj4m7nj5jm-fortify-headers-1.1alpine1/include
 /nix/store/hcv14gpi8ja7san8sxpkqrrqdgn8a7yn-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-musl/12.2.0/include-fixed
End of search list.
```

Reply:
I don't really understand this because the gcc docs say the includes are visited from left to right.

Reply:
`ignoring duplicate directory "/nix/store/pd1dj9hbndmhndv34yffnr3m7dy4503p-musl-1.2.3-dev/include"` might be a hint.

I suspect it comes from default `gcc`'s embedded path set by `--with-native-system-header-dir=`. I've been meaning to remove that from default search paths as well, but did not come up yet with a clean way to do it yet :(

Reply:
Filed https://github.com/NixOS/nixpkgs/issues/245455 for it. I don't plan to address it myself any time soon, but I think it's a very important bug.